filters/ema.cpp

```diff
@@ -1,24 +1,144 @@
+#include <utility>
+
 #include "filters/ema.hpp"
 
+#include "serenity/metrics_helper.hpp"
+
 namespace mesos {
 namespace serenity {
 
-IpcEMAFilter::~IpcEMAFilter() {}
+double_t ExponentialMovingAverage::calculateEMA(
+    double_t sample, double_t sampleTimestamp) {
+  if (this->uninitialized) {
+    this->prevEma = sample;
+    this->uninitialized = false;
+  }
+
+  switch (seriesType) {
+    case EMA_REGULAR_SERIES:
+      this->prevEma = this->exponentialMovingAverageRegular(sample);
+      break;
+    case EMA_IRRERGULAR_SERIES:
+      // TODO(bplotka): Test irregular series EMA.
+      this->prevEma = this->exponentialMovingAverageIrregular(
+          sample, sampleTimestamp);
+      break;
+  }
+
+  this->prevSample = sample;
+  this->prevSampleTimestamp = sampleTimestamp;
+
+  return prevEma;
+}
+
+
+double_t ExponentialMovingAverage::exponentialMovingAverageIrregular(
+    double_t sample, double_t sampleTimestamp) const {
+  double_t deltaTime = sampleTimestamp - this->prevSampleTimestamp;
+  double_t dynamicAlpha = deltaTime / this->alpha;
+  double_t weight = exp(dynamicAlpha * -1);
+  double_t dynamicWeight = (1 - weight) / dynamicAlpha;
+  return (weight * this->prevEma) + (( dynamicWeight - weight )
+          * this->prevSample) + ((1.0 - dynamicWeight) * sample);
+}
+
+
+double_t ExponentialMovingAverage::exponentialMovingAverageRegular(
+    double_t sample) const {
+  return (this->alpha * sample) + ((1 - this->alpha) * this->prevEma);
+}
+
+
+Try<Nothing> EMAFilter::consume(const ResourceUsage& in) {
+  std::unique_ptr<ExecutorSet> newSamples(new ExecutorSet());
+  ResourceUsage product;
+
+  for (ResourceUsage_Executor inExec : in.executors()) {
+    if (!inExec.has_statistics() || !inExec.has_executor_info()) {
```

Comment:
@ConnorDoyle: Good question! We can either do it here to calculate EMA only for PR jobs, but it would be better to have it filtered out earlier as @skonefal said. First option will be better if we need BE executors for some calculations in any component in pipeline. So far.. we do not.